### PR TITLE
Reward programs agent

### PR DESCRIPTION
**Pull request by dgusakov**
Add lido-rewards agent
https://www.notion.so/Rewards-accounts-balance-monitoring-3dfa2d0507de48ae91551cb35193f33e

## Description
We have 5 000 000 LDO cap to spend on reward programs each month.
LDO can be allocated to reward programs only throw the top-up easy-track motions.
We need to have up-to-date info about LDO spend on rewards this month, so we can be confident about the absence of underspending and overspending.

### How agent works
Agent checks each TX for MotionEnacted and MotionCreated events related to the TopUpRewards motion type.
- On MotionCreated event, amount of LDO is added to "pending".
- On MotionEnacted event, amount of LDO is removed from "pending" and added to "spent".
- On the first block in new month "spent" amount is set to 0.
- On agent start, agent fetches blockchain history up to the start of the month so that we have full info related to this month

### CODE REVIEW

**Review by skhomuti — APPROVED** [review]

File: lido-rewards/src/agent-reward-programs.ts [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+      if (
+        creationEvent &&
+        creationEvent.args &&
+        creationEvent.args._evmScriptFactory.toLowerCase() ==
+          TOP_UP_REWARDS_ADDRESS
+      ) {
+        const callData = await topUpRewards.functions.decodeEVMScriptCallData(
+          creationEvent.args._evmScriptCallData
+        );
+        let sum_amount = new BigNumber(0);
```
Comment (vgorkavenko) [inline lido-rewards/src/agent-reward-programs.ts:82]:
```suggestion
        let sumAmount = 0n;
```

File: lido-rewards/src/agent-reward-programs.ts [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        callData._amounts.map((value: any) => {
+          sum_amount = sum_amount.plus(new BigNumber(String(value)));
```
Comment (vgorkavenko) [inline lido-rewards/src/agent-reward-programs.ts:84]:
```suggestion
          sumAmount = sumAmount + BigInt(String(value));
```

File: lido-rewards/src/agent-reward-programs.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    motionsCreated.map(async (value: any) => {
+      const id = parseInt(String(value.args._motionId));
+      if (
+        value.args._evmScriptFactory.toLowerCase() == TOP_UP_REWARDS_ADDRESS &&
+        !enactedTopUpMotions.get(id)
+      ) {
+        const callData = await topUpRewards.functions.decodeEVMScriptCallData(
+          value.args._evmScriptCallData
+        );
+        let sum_amount = new BigNumber(0);
```
Comment (vgorkavenko) [inline lido-rewards/src/agent-reward-programs.ts:112]: the same as above

File: lido-rewards/src/agent-reward-programs.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    )} LDO`
+  );
+  return {
+    pendingMotions: `[${Array.from(pendingTopUpMotions.keys()).join(",")}]`,
+    enactedMotions: `[${Array.from(enactedTopUpMotions.keys()).join(",")}]`,
+  };
+}
+
+function sumMapValues(mapping: Map<number, BigNumber>) {
+  let sum = new BigNumber(0);
```
Comment (vgorkavenko) [inline lido-rewards/src/agent-reward-programs.ts:140]: the same as above

File: lido-rewards/src/agent-reward-programs.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    if (eventCreated) {
+      const id = parseInt(String(eventCreated.args._motionId));
+      if (
+        eventCreated.args._evmScriptFactory.toLowerCase() ==
+        TOP_UP_REWARDS_ADDRESS
+      ) {
+        const callData = await topUpRewards.functions.decodeEVMScriptCallData(
+          eventCreated.args._evmScriptCallData
+        );
+        let sum_amount = new BigNumber(0);
```
Comment (vgorkavenko) [inline lido-rewards/src/agent-reward-programs.ts:202]: the same as above

**Review by vgorkavenko — APPROVED** [review]

File: lido-rewards/src/constants.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import BigNumber from 'bignumber.js'
+
+
+// COMMON CONSTS
+
+// 1 LDO
+export const LDO_DECIMALS = new BigNumber(10).pow(18)
```
Comment (madlabman) [inline lido-rewards/src/constants.ts:7]: As for me it's confusing to have `10^18` in `LDO_DECIMALS` instead of 18

**Pull request merged**